{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Outline for Python Scripting for Biochemists\n",
    "This workshop is based on Python Scripting for [Computational Molecular Sciences](http://education.molssi.org/python_scripting_cms/)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modules\n",
    "1. Setup\n",
    "2. Introduction with calculations using enzyme kinetics\n",
    "3. File parsing for data sets for a promiscuous enzyme with multiple substrates\n",
    "4. Plotting and linear curve fitting for protein concentration calibration curves using numpy\n",
    "5. Plotting and nonlinear curve fitting for saturable single substrate enzyme kinetics using numpy\n",
    "6. Writing functions\n",
    "7. Running code from the Linux command line\n",
    "8. Testing code with pytest\n",
    "9. Version control with git \n",
    "10. Sharing code\n",
    "11. Introduction to panda using data sets derived from the PDB\n",
    "12. Plotting UV-Visible spectra using matplotlib including making interactive spectra with magic % "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Meeting Notes: BASIL team on Monday, 25 Jan 2021\n",
    "<!--- ![BASIL Team Interview](BASIL_Interview_Monday_25Jan2021.jpg)-->\n",
    "\n",
    "Date: Monday January 25, 2021\n",
    "Interview with BASIL Team members: Bonnie Hall, Julia Koeppe, Jose Tormos Melendez, Cassidy Terrell\n",
    "\n",
    "### Details\n",
    "I told them that I was creating a workshop based on Ashley's Python Scripting for CMS workshops and that I wanted to focus on things biochemists need to learn.\n",
    "1. Manipulating a PDB file\n",
    "    1. Using Biopython\n",
    "    1. What data would you want to extract from a PDB file?\n",
    "    1. They might want to measure a distance\n",
    "1. Enzyme kinetics from a plate reader\n",
    "    1. lots of data\n",
    "    1. merging multiple data sets\n",
    "    1. converting raw data to derivatized data\n",
    "        1. students don't understand that it is a time course for absorbance values as a function of time at a fixed wavelength\n",
    "        1. subtracting a negative control\n",
    "    1. look over the data and decide to reduce or increase enzyme concentration\n",
    "1. Tools that would be useful\n",
    "    1. LaTeX equations training\n",
    "    1. Greek letters\n",
    "    1. Quadratic equations\n",
    "    1. Equilibrium calculations: e.g., is this ionizable group protonated or deprotonated at a given pH?\n",
    "    1. pH profiles of activities for enzymes with amino acid substitutions\n",
    "        1. Does the activity change?\n",
    "        1. Does the pH optimum change?\n",
    "        1. Can you extract a pK<sub>a</sub> value from the data?\n",
    "1. More sophisticated ideas\n",
    "    1. Ask students to use python to reproduce a plot given the data \n",
    "    1. Predictive modeling and clustering\n",
    "    1. K nearest neighbor\n",
    "    1. Heat mapping data\n",
    "    \n",
    "### Emerging ideas\n",
    "- Students need to learn things they will use\n",
    "- A toolbox of resources would be nice\n",
    "    - Greek letters\n",
    "    - LaTeX equation training\n",
    "    - Common calculations: pH, pKa, M-M, quadratic equations\n",
    "- Plate readers produce lots of data. Tools for merging, manipulating, and analyzing these data files would be useful\n",
    "    - enzyme assays\n",
    "    - protein assays\n",
    "    - immunoassays\n",
    "- Some faculty will want more advanced tools for their research students\n",
    "    - Heat mapping (what is that?)\n",
    "    - K nearest neighbor analysis\n",
    "    - Predictive modeling and clustering\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Meeting Notes: Phil Ortiz on Wednesday, 27 Jan 2021\n",
    "<!--- ![Phil Ortiz Interview](Phil_Ortiz_Interview_Wednesday_27Jan2021.jpg)-->\n",
    "\n",
    "Date: Wednesday January 27, 2021\n",
    "Interview with Phil Ortiz\n",
    "\n",
    "### General ideas\n",
    "1. He is not certain that students need any coding experience. \n",
    "1. He thinks they definitely need to know how to use a spreadsheet to analyze and plot data.\n",
    "    1. He strongly favors having students learn to build their own spreadsheets\n",
    "    1. He learned on Lotus 1-2-3 and Sigma Plot\n",
    "1. He is a strong advocate of quantitative skills for the students\n",
    "1. He mentioned a number of challenging calculations\n",
    "    1. Michaelis-Menten\n",
    "    1. Henderson-Hasselbalch: pH, pK_a\n",
    "1. Computation could be handy for some more advanced topics\n",
    "    1. Structure-function relationships: how does changing pH change the function of a protein?\n",
    "    1. Hemoglobin would be a great example for this\n",
    "1. Plots are important and difficult to generate\n",
    "    1. Hydropathy plots\n",
    "    1. Plot of helical wheels with colors to indidate charge and polarity (even better in 3D)\n",
    "    1. It might be useful to go through a biochem textbook and look at the plots that are there\n",
    "\n",
    "### Emerging Ideas\n",
    "- Use of existing software (Excel) is one level of computing skill that should be reinforced\n",
    "- Some tools for computation would be handy\n",
    "- It would be worth reviewing all the plots in Tansey's text to see which ones can be done in Excel and which need Python or something more."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Meeting Notes: Ashley Ringer McDonald on Friday, 29 Jan 2021\n",
    "Date: 29 January 2021\n",
    "### General Ideas\n",
    "1. When considering the Python Scripting for CMS, different is better\n",
    "    1. The file parsing lesson could focus on a larger PDB file\n",
    "        1. Take a PDB file and write a new file with only the title, ATOM and HETATM records\n",
    "        1. Selectively remove ANISOU entries\n",
    "        1. Selectively extract only x, y, z coordinates\n",
    "        1. Look for specific ligands\n",
    "    1. The pandas library could be introduced\n",
    "    1. Regular expression matching might be more important\n",
    "1. Use the SciPy library for curve-fitting. It has better syntax than NumPy\n",
    "1. Do we have any need for symbolic math? I told her no, since I don't know what to do with it\n",
    "1. She recommended reviewing the MolSSI [Python for Data Analysis workshop](https://education.molssi.org/python-data-analysis/)\n",
    "1. She recommended getting workshops onto venues sooner rather than later\n",
    "    1. ASBMB Teaching for the Molecular Biosciences\n",
    "    1. Better luck at regional and specialty meetings\n",
    "        1. ACS-regional(NERM)\n",
    "        1. Rochester ACS section\n",
    "        1. local institutions\n",
    "    1. Summer research programs\n",
    "    1. NIH Biophysics training grant\n",
    "    1. Teach the teachers workshop for biochemists\n",
    "    \n",
    "### Emerging Ideas\n",
    "- Start working on workshop planning\n",
    "- Don't be afraid to take a different path than the Python Scripting for CMS. Different is better.\n",
    "- Consider other libraries: Pandas, SciPy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Meeting Notes: Wally Novak on Monday, 1 Feb 2021\n",
    "Date: Monday 1 Feb 2021\n",
    "### General Ideas\n",
    "1. The conversation focused a lot on pedagogy - how and what we want to teach. We had a good discussion about creating tools that will support hypothesis generation based on \n",
    "    1. sequence comparison\n",
    "    1. structure comparison\n",
    "1. The sequence comparison should go well beyond what can be done in BLAST and Pfam\n",
    "    1. There are so many identical sequences it is difficult to pull useful data from those sites any more.\n",
    "    1. It is more powerful from the command line, where parameters are more readily controlled.\n",
    "    1. Cluster analysis is powerful and possible\n",
    "1. Another challenge is correlating sequence and structural data\n",
    "    1. Identifying how closely residues from crystals align. This is especially interesting in comparing cases like subtilisin and trypsin.\n",
    "    1. Follow these comparisons with sequence alignments\n",
    "    1. Learn to wrtie python codes to look for patterns of residues in the entire PDB\n",
    "1. Emphasis on two questions when approaching programming:\n",
    "    1. How it works (not so important for BMB students and faculty)\n",
    "    1. What it can do for you (very important)\n",
    "1. He also mentioned learning about and using Biopython\n",
    "\n",
    "\n",
    "### Emerging Ideas\n",
    "- Need to consider correlating sequence and structural alignments\n",
    "- How much command line instruction do we want to provide?\n",
    "- Should we consider free-standing Python or focus only on Jupyter notebooks\n",
    "* Overall we must emphasize \"what can it do for me?\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Meeting Notes: Charlie Weiss on Monday, 1 Feb 2021\n",
    "\n",
    "### General Ideas\n",
    "Our conversation focused mainly on libraries and skills. Charlie is an organic and inorganic chemist.\n",
    "1. Libraries\n",
    "    1. Numpy is powerful but basic\n",
    "    1. Pandas has more features and datatypes (e.g. dataframes) for data importing. His students love pandas.\n",
    "    1. SciPy is for plotting. I need to read the chapter in his book.\n",
    "        1. built on numpy\n",
    "        1. basic\n",
    "        1. good for integration\n",
    "        1. Signal processing\n",
    "        1. Fourier transform and reverse Fourier transform\n",
    "    1. Visualization libraries\n",
    "        1. Matplotlib is powerful but basic and verbose\n",
    "        1. Seaborn is built on matplotlib. You can make a complex plot with 1-2 lines. See Charlie's book.\n",
    "    1. Biopython\n",
    "        1. stable and pretty old\n",
    "        1. many for loops required\n",
    "        1. good for bringing in data repeatedly or with similar patterns\n",
    "    1. SciKitBio\n",
    "        1. in flux - still listed as in beta\n",
    "        1. you can use it to poke around at data\n",
    "        1. It wil convert DNA sequences to RNA and protein sequences\n",
    "    1. SciKitLearn\n",
    "    1. NMRGlue\n",
    "1. I mentioned my desire to bring in 3D visualization.\n",
    "    1. He was not aware of any molecular viewers for Jupyter notebooks.\n",
    "    1. He said that the Jupyter developers are very friendly and responsive.\n",
    "    1. He recommended two conferences\n",
    "        1. Jupyter Con\n",
    "        1. Annual SciPy conference in Austin, TX\n",
    "            1. Some presentations\n",
    "            1. Lots of impromptu discussions\n",
    "            1. Birds of a feather sessions were really helpful\n",
    "1. Skills\n",
    "    1. Dealing with large datasets that can't be managed on a local computer\n",
    "    1. Out of core methods (programs that must be run on a supercomputer?)\n",
    "\n",
    "### Emerging Ideas\n",
    "- Some of these libraries should be included in the workshop: numpy, scipy, pandas\n",
    "- There are additional libraries that should be part of followup materials\n",
    "- I need to get to JupyterCon and the SciPy meeting in Austin, TX"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Meeting Notes: Dan Dries on Tuesday, 9 Feb 2021\n",
    "\n",
    "### General Ideas\n",
    "\n",
    "### Emerging Ideas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Meeting Notes: Joe Provost\n",
    "### General Ideas\n",
    "\n",
    "### Emerging Ideas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Meeting Notes: Bonnie Hall on Wednesday, 4 Feb 2021\n",
    "### General Ideas\n",
    "1. Start with simple linear regression\n",
    "    1. Use a common library - matplotlib, numpy or scipy\n",
    "    1. Base it on a common target - Bradford protein assay or SDS-PAGE mobility\n",
    "    1. Take data from Excel -> CSV -> Jupyter/Python\n",
    "    1. At some point we may want to recruit someone to translate this to R\n",
    "    1. Fit the curve (slope, intersect, R-squared), then graph the curve\n",
    "    1. Include error\n",
    "        1. In replicate points\n",
    "        1. For the curve itself - where is is most and least reliable?\n",
    "1. Predictive modeling\n",
    "    1. Classification\n",
    "        1. K-nearest neighbor\n",
    "        1. Many data points\n",
    "        1. Sort the information by proximity\n",
    "    1. Examples\n",
    "        1. Protein sequence\n",
    "        1. Protein function, e.g., hydrolases with substrate results\n",
    "1. Kinetics calculations using output from a plate reader\n",
    "    1. Absorbance vs. time to get an initial slope\n",
    "        1. Use this to estimate substrate consumption\n",
    "        1. Is it linear? If not, how should the experiment be changed?\n",
    "    1. Use this to calculate the activity\n",
    "    1. An exercise to handle tedious math\n",
    "1. Information from the PDB\n",
    "1. Data wrangling using Pandas (including the cheat sheet Bonnie provided)\n",
    "    1. Take data from a .csv file but don't modify the original file\n",
    "    1. Run code that does calculations and analysis and output to a new file\n",
    "    1. Search for useful points: minima, maxima, null values\n",
    "    1. Change ligand that are the same but have different names\n",
    "    1. One possible data source is metabolomics raw data\n",
    "\n",
    "### Emerging Ideas\n",
    "- Simple is good. Simple linear regression will be well received and make them comfortable.\n",
    "- Very interested in data wrangling with pandas\n",
    "- Put useful tools in their hands that they and their students will use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
